### ISSUE TITLE
Unable to sync - an error occurred

### ISSUE THREAD

**Issue by scooterlord** [10up/ElasticPress#2083]
Hello,
I am getting this in the latest Elasticpress - 3.5.4 and php 7.414.

As has been suggested in a closed thread in this forum (https://github.com/10up/ElasticPress/issues/924#issuecomment-778218299) I have done the following:

- Increased memory limit to 1gb
- Increased PHP timeout to 60s

During syncing, I get 'Unable to syn an error occurred' at random times - console indicates `500 error` and it just stops. I tried with various settings for the Items per cycle - 100,200,350,500.. I can see a huge speed differences for different settings but the result remains the same - an error is occurred. It's trying to sync something like 21000 items.

As has been suggested by @felipeelia I used `wp-cli ` and, although I get 'Warning: too many requests' several times, the procedure completes successfully - which essentially makes me wonder that whenever a warning occurs in the front-end the whole procedure just halts.

Suggestions for improvement EXCEPT solving the bug - maybe these deserve their own thread, but thought I'd drop them here for convenience.
- When anything happens while syncing - either good or bad - the message disappears after a few seconds, which essentially means that you have to sit on your screen while doing the sync - I am certain there are other ways to review it - maybe by examining logs, etc - but... why? Please make the message stay on screen.
- More importantly, in such cases, why halt the procedure completely? Adding an action like 'Too many requests: Retry?` would be a lot better.
- To take things one step further you could provide some additional details upon completion, like a small table with stats like Total sync time, sync time per item etc, so that people can evaluate which option works best for them on their environments.

Thank you for this wonderful plugin, I hope it stays alive and keep improving over time!

**Comment by joliverwestbrook** [comment]
I've also been seeing this exact problem. In our case I bypassed the generic 500 error that spits out from `includes/dashboard.php` around line 628, and instead I drop the return into a log file. This works in logging the errors (and successes) so far, until I put my chunks limit at 100 (I started with 1), and then got a "legit" 500 error.  I also saw a lot of times, when navigating to the ElasticPress settings, the notice at the top that it could not reach the elasticpress.io servers.

So here are my findings so far. I'm going to fix this for our site, but would love if a fix can be done for the plugin so we can keep up with updates, as my fix will mean we cannot update or we end up breaking sync again.

About our environment:
- ~953 posts in total to sync
- Custom meta attributes for one of our CPTs (we have 5 cpts)
- WP 5.6.1
- PHP 7.3.27
    >> Changed WP cURL to 500 timeout - this helped, but did not resolve all errors
    >> Increased PHP timeouts, did not seem to effect
- EP 3.5.3 (I've tried the latest version also and it has the same results)
- Single-site

So far it seems increasing the wordpress curl timeout helped in eliminating some of the sort of "random" errors I was seeing. However, like I said, if I increase the sync items above 1, it starts to just have a generic 500 error, regardless of my edited dashboard file with custom error logging.

And, the other error I now see, at random times during a sync of 1 item per cycle, is: 
`Limit of total fields [1000] in index [xxxxxxxxxxxxxxxxxxxx-post-1] has been exceeded`

As stated, we are only presently trying to sync about 953 items.  Is there a way to "refresh" our index or start over with a fresh sync? Is that a possible solution at this point?  What is this 1000 limit?

**Comment by joliverwestbrook** [comment]
One more note on my findings:

It does seem to have errors and fail only with posts from a single CPT. If I trash the posts it runs fine. What this has lead me to believe is some of the ~800 posts in that CPT may be effectively "corrupted" in terms of how this data is handled during a sync. Is this a known possibility? Or is it more likely these posts have more data in general and are causing the failer? Does 1000 limit I see in these single sync errors refer to 1000 items within an index (1000 posts) or 1000 items within a given posts meta?

**Comment by joliverwestbrook** [comment]
I found and changed the 1000 limit, to be 5000 per shard by setting it to 25000 on line 38 of `/includes/mappings/post/7-0.php` - test is still running, will report the results.  This includes the curl change to 500 timeout and running a sync with a cycle setting of 1

**Comment by scooterlord** [comment]
FYI, I am syncing default articles - not CPTs.

**Comment by joliverwestbrook** [comment]
Okay, so I was able to fix this issue by performing the following steps:

1. Increase PHP Mem and Post limits to 1024M
2. Add faster nameservers (google and cloudflare) to the server resolv file
2. Increase WP cURL timeout to 300 (5 min)
2. Increase WP mem limit to 1024M in wp-config
3. Limit my cycle setting to 50 in EP settings
4. Update to latest EP (3.5.4)
5. Edited EP file `/includes/mappings/post/7-0.php` at line 22, changing the value to `25` (this is the shard count limit)
6. Edited EP file `/includes/mappings/post/7-0.php` at line 38, changing the value to `25000` (this is the max index limit, divided by shard count, 1K per shard...but allowing more shards)

None of these fixes applied individually worked consistently, all of them together however have yet to produce a failure or error.  In addition, to help debug I altered the section of code in `/includes/dashboard.php` to output errors to a log rather than issue a 500 header and fail, which ends up killing the entire sync and producing no readable errors.

**Comment by scooterlord** [comment]
@joliverwestbrook Hello again and thanks for your support.

Your posts also validate the inconsistency I saw also, I have a few questions regarding your last comment:

- Is the PHP limit different than the WP memory limit? I thought it was either this or that...?
- Where did you change the WP cURL timeout?
- More or less I know what shards do, but what is the connection to specific wordpress entities? 
- Limiting to 50 items per cycle, didn't make things very slow for you?

As for myself, I was already using the 1gb limit through wp-config, already using cloudflare nameservers and was on latest EP version.

Editing the files is a no-go for me. I've done it in the past, but I have a million things on my mind and tens of websites to maintain - although you can override the limit I guess using a function. Regarding the limit, I see no benefit for anyone to have a limit. I guess one would want to index everything and either they are below or above this limit.

As a sidenote, I am also using a caching plugin, so it was necessary to clear the cache after changing the items-per-cycle option to use the new setting.

**Comment by scooterlord** [comment]
Today, I switched to a Digital Ocean server and encountering the same problem. Unfortunately, I was not able to successfully install/run wp-cli, so I am stuck with errors. I changed wordpress memory limit and php memory limit to 2048mb. still I am getting errors. 

I would expect the procedure to proceed as in the cli version, aka, resume if for any reason it fails temporarily.

@joliverwestbrook Any more help you can provide me with?

**Comment by felipeelia** [comment]
Hi @scooterlord,

Those three suggestions you made in the first message would be excellent additions to the plugin, Do you mind opening three new issues with the "enhancement" type?

Regarding the `Limit of total fields [1000] in index [xxxxxxxxxxxxxxxxxxxx-post-1] has been exceeded` error mentioned by @joliverwestbrook, we have this covered in this article: https://elasticpress.zendesk.com/hc/en-us/articles/360051401212-I-get-the-error-Limit-of-total-fields-in-index-has-been-exceeded-
Basically, it is related to the total number of meta fields present in the site and how ElasticPress manipulates each one of them.

Can you both please give a look into that article, make the necessary checks and let us know if it helps to solve the issue? Thanks!

**Comment by scooterlord** [comment]
@felipeelia I have created a new ticket with my suggestions.

Going back to the initial issue I am facing, in my current (new) installation, for some reason, although I managed to install wp-cli, when running 'wp elasticpress' commands I get an 'Error: could not connect to Elasticsearch.' which I have no idea how to search for - wp-cli seems to be working since I have access to other wordpress commands. So, in essence I am stuck. I apologize for semi-hijacking this thread but if there are any suggestions, please share.

I don't have the fields limit issue as mentioned above.

**Comment by scooterlord** [comment]
Hello @felipeelia and thanks for your interest in my problem. My site is a direct replication of the same app on a different server, settings are exactly the same. Wordpress itself can access my EP server, just not when accessing via ssh through wp-cli!

**Comment by joliverwestbrook** [comment]
> Hi @scooterlord,
> 
> Those three suggestions you made in the first message would be excellent additions to the plugin, Do you mind opening three new issues with the "enhancement" type?
> 
> Regarding the `Limit of total fields [1000] in index [xxxxxxxxxxxxxxxxxxxx-post-1] has been exceeded` error mentioned by @joliverwestbrook, we have this covered in this article: https://elasticpress.zendesk.com/hc/en-us/articles/360051401212-I-get-the-error-Limit-of-total-fields-in-index-has-been-exceeded-
> Basically, it is related to the total number of meta fields present in the site and how ElasticPress manipulates each one of them.
> 
> Can you both please give a look into that article, make the necessary checks and let us know if it helps to solve the issue? Thanks!

I can verify this. After further work we discovered our server had an enormous amount of duplicate postmeta. I have a very rough function I wrote to find dupe-meta and remove it across posttypes you choose: https://github.com/joliverwestbrook/WP-CleanDupeMeta/blob/dev/simple-postmeta-dedupe.php

After running this, I was able to remove the custom code changes made to the mappings settings in the 7-0.php file and the sync was able to run correctly.

**Comment by joliverwestbrook** [comment]
> Hello @felipeelia and thanks for your interest in my problem. My site is a direct replication of the same app on a different server, settings are exactly the same. Wordpress itself can access my EP server, just not when accessing via ssh through wp-cli!

Sorry for the late reply. So the can't connect issue is one I was seeing also, when the size of the data was just too big (due to the dupe postmeta issue we found). This could cause a timeout of sorts that could generate a connection error, even though it's only a connection error due to the other issues at hand.

To answer your Q's from earlier:
- PHP and WP mem settings are the same, WP settings override php.ini settings.
- cURL timeout can be set in your functions file using two hooks, as the coding I did is for a client I'm not at liberty to post here, but a simple search for something like "increase wp curl timeout" should get you a good answer.
- I don't really fully know how to explain shards in this context, I was just using logic to understand the relation between shards and items.
- 50 limit did not make a significant difference in time to sync.  After removing postmeta I can take it to 100 without issue, but I did have one run at that level which failed...however, the failure that time did not make me have to star the sync over entirely, when I hit sync it picked up where it left off.  So the error before generated a 500, which I believe is why it didn't pick up where it left off when I tried again at that time.

Hope that's helpful and good luck fixing yours!

**Comment by scooterlord** [comment]
Hello again and thanks for you replies. My problem in the new server is that although everything is setup like the previous server the sync halts near the beginning. I have a lot of meta but not THAT many like maybe 20-30 per post. I set memory to 2gb, increased curl time (found the filters online), but still the problems remain.

On top of that in the new server (digital ocean) I encountered lots of problems trying to setup wp-cli, which produces the error that can’t connect to elastic search as soon as it runs. If I managed to make it work via cli I wouldn’t care, the problem is that the error message is too generic and don’t know how to investigate.

**Comment by scooterlord** [comment]
Today that I tried the same function on my digital ocean it just worked! I am pasting a screenshot that shows the number of errors that occur, however, the syncing appears to be completed normally despite the errors - the thing is that even if an error occurs the process still proceeds, which is what I would expect from the front-end to to as well.

![image](https://user-images.githubusercontent.com/5257454/108680046-06375b00-74f6-11eb-97ab-5736d60ca56a.png)

**Comment by scooterlord** [comment]
..Also reporting that the `--no-bulk` option worked as expected without errors, but really for 22000 items being synced it was a pain to watch. During this sync, my ES node instance showed RAM utilization 80-93% (1GB Ram). Next time I will try with 50 items per cycle and increase, but still, makes me wonder how come the same data from a different server sync properly with 350 items per cycle on the same target node.

**Comment by joliverwestbrook** [comment]
@scooterlord have you tried configuring a new droplet NOT like the other server? What I mean is, try a different config altogether to rule out your config or some part of it being at the root of the issue.

Another thing, from what region are you connecting to EP's servers and is it close in proximity? This shouldn't matter really, but to trace the issue down you'll need a very disciplined debugging strategy and controlled environment. When you get a server that works for example, clone it and see if the clone works and if so, begin making small changes and tests, see if you can reverse it and create the problem from a working server...and vise versa.

The other thing I would suggest is to see if you can narrow down the issue via post types and even posts, by trashing all posts of a certain type, then run a sync, see if it struggles, etc., see if it's a particular post type which, on the back end may have more meta data or perhaps "specialized" meta data that could be interfering with the sync process.

**Comment by scooterlord** [comment]
@joliverwestbrook I wasn't replicating an existing droplet. I created a new plesk droplet and replicated files/db from a completely different environment. Everything else seems to be working as expected.

As I said in my previous reply, I only synced posts  - this would be a great suggestion to the devs btw, select what kinds of post types to sync. Trying in bulks of 50s still gave errors. :/

The thing is that it doesn't consistently occur at a specific post, it appears to happen at random. The 'no-bulk' option seems to be working fine, although I am not really sure it will help in the long run - for now it's _just_ 4500 posts, but the plan is that it will get a lot larger. Fingers crossed I guess. :/

**Comment by joliverwestbrook** [comment]
That is really odd. I wish I had more to offer but best of luck in debugging.

**Comment by scooterlord** [comment]
Continuing, today I contacted the support of my elasticsearch service. I am using the smallest package in elastic.co which I think that it's struggling - only when indexing but this is still a problem. Does anyone have any good alternatives to suggest?

**Comment by brandwaffle** [comment]
@scooterlord I'm pretty sure I've mentioned ElasticPress.io to you before :)

In all seriousness, you're running into the exact problems we had when we tried to run Elasticsearch on services like Elastic.co but for WordPress search (a pretty unique use-case that differs from a lot of the largest Elasticsearch installs out there), which is why we built our custom service that includes all the features you're going to need and is guaranteed to be fully plugin-compatible.

Given that you've seen the low-end Elastic.co plan is insufficient, and if you factor in the time you've spent chasing down an infrastructure issue, is there a price point you feel would be reasonable for a service that "just works"? As we expand, we're constantly thinking about ways to deliver more value at lower cost, so I would appreciate your input as to where that value point would be for you.

**Comment by scooterlord** [comment]
@brandwaffle thank you for your reply. I had evaluated your service in the past and what was very appealing was that - if I remember correctly - you provided autosuggest functionality out of the box.

In all honesty though your service is too expensive and too limited - especially when defining number of records etc. so in our case, now that we just started we have 4500 posts, this time next year we’ll have probably 10x. So, your cheapest plan, which is still 4x times of the plan I’m using in elastic.co, won’t be enough for our project. We just started, our business model is free tier for now and we have expenses all over the place, elastic is just added to other expenses.

If I felt more confident I’d setup elastic on a digital ocean server and work from there, but since I am not is the reason I am looking for a managed service. I am developing a huge application on my own, so elasticsearch is just 1-20th of what I have to deal with - UX, design, Wordpress, etc..

I can’t imagine I am alone in this situation, and it’s good to hear that you are evaluating other cheaper options for audiences like me. You will propably feel it’s too low for your services based on your current pricing but would be willing to pay up to 40eur/mo for a plug-and-play solution without limitations. I mean in the end, it’s just a system, why limit records? You probably have your own policies, but being more generous and giving greater flexibility will get you more income in the long run.

**Comment by felipeelia** [comment]
That is excellent news @scooterlord . Thank you *very much* for coming back here and sharing your findings. I'm sure it'll help any future readers of this issue!

I'm closing it right now, but feel free to open a new one if you face any other different problem, okay? Thanks!

**Comment by brandwaffle** [comment]
@scooterlord I just wanted to chime in and thank you for your feedback! We're actively working on plans that don't include our ticketed free support (similar to a typical cloud offering but optimized for WordPress/ElasticPress and including our custom ElasticPress.io feature set), at price points within (or below) the 40euro/month range.

**Comment by scooterlord** [comment]
@felipeelia @brandwaffle thank you guys! 

@brandwaffle, this is great news! I would gladly give it a spin - the autocomplete feature alone makes it worth it for me - currently we are managing without it. And please, be _elastic_ with those limits/restrictions! I can only assume that there's a lot of room for profit, so...

Looking forward to what you guess come up with!
